contracts/Turnstone.vy

```diff
@@ -0,0 +1,108 @@
+
+last_checkpoint: public(bytes32)
+last_valset_nonce: public(uint256)
+
+@external
+def __init__(turnstone_id: bytes32, validators: DynArray[address, MAX_VALIDATORS], powers: DynArray[uint256, MAX_VALIDATORS]):
+    TURNSTONE_ID = turnstone_id
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in validators:
+        cumulative_power += powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    valset_args: ValsetArgs = ValsetArgs({validators: validators, powers: powers, nonce: 0})
+    self.last_valset_nonce = 0
+    new_checkpoint: bytes32 = keccak256(_abi_encode(valset_args.validators, valset_args.powers, method_id=method_id("checkpoint(address[],uint256[])")))
+    self.last_checkpoint = new_checkpoint
+
+@external
+@pure
+def turnstone_id() -> bytes32:
+    return TURNSTONE_ID
+
+@internal
+@pure
+def verify_signature(signer: address, hash: bytes32, sig: Signature) -> bool:
+    message_digest: bytes32 = keccak256(concat(convert("\x19EthereumSignedMessage:\n32", Bytes[26]), hash))
+    return signer == ecrecover(message_digest, sig.v, sig.r, sig.s)
+
+@internal
+def check_validator_signatures(current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS], hash: bytes32):
+    i: uint256 = 0
+    cumulative_power: uint256 = 0
+    for sig in sigs:
+        if sig.v != 0:
+            assert self.verify_signature(current_valset.validators[i], hash, sig), "Invalid Signature"
+            cumulative_power += current_valset.powers[i]
+            if cumulative_power >= POWER_THRESHOLD:
+                break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+
+@internal
+@view
+def make_checkpoint(valset_args: ValsetArgs) -> bytes32:
+    return keccak256(_abi_encode(valset_args.validators, valset_args.powers, valset_args.nonce, TURNSTONE_ID, method_id=method_id("checkpoint(address[],uint256[],uint256,uint256)")))
+
+@external
+def update_valset(new_valset: ValsetArgs, current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS]):
+    assert new_valset.nonce > current_valset.nonce, "Invalid Valset Nonce"
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in new_valset.validators:
+        cumulative_power += new_valset.powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    assert self.last_checkpoint == self.make_checkpoint(current_valset), "Incorrect Checkpoint"
+    new_checkpoint: bytes32 = self.make_checkpoint(new_valset)
+    self.check_validator_signatures(current_valset, sigs, new_checkpoint)
+    self.last_checkpoint = new_checkpoint
+
+@external
+def submit_logic_call(current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS], args: LogicCallArgs, message_id: uint256, deadline: uint256):
```

Comment:
Given that all function calls originating from paloma will always include ValsetArgs and sigs, what do you say that we create a class that would represent just that. It would make this nicer to play with and it would make much more sense who is supposed to be calling the provided function just by looking at the function signature. Proposed class:

```python
class Consensus:
  validators: ValsetArgs
  signatures: DynArray[Signature, MAX_VALIDATORS]
```

What do you think?